sound/soc/sof/sof-client-probes.c

```diff
@@ -188,6 +194,56 @@ static const struct snd_compress_ops sof_probes_compressed_ops = {
 	.copy = sof_probes_compr_copy,
 };
 
+#ifdef CONFIG_SND_SOC_SOF_IPC4
+static const char *sof_probe_ipc4_type(u32 type)
+{
+	switch (type) {
+	case PROBE_TYPE_INPUT:
+		return "input";
+	case PROBE_TYPE_OUTPUT:
+		return "output";
+	case PROBE_TYPE_INTERNAL:
+		return "internal";
+	default:
+		return "UNKNOWN";
+	}
+}
+
+static int sof_probes_ipc4_prints(struct sof_client_dev *cdev, char *buf, size_t size,
+				  struct sof_probe_point_desc *desc)
+{
+	struct snd_sof_dev *sdev = sof_client_dev_to_sof_dev(cdev);
+	struct device *dev = &cdev->auxdev.dev;
+	struct snd_sof_widget *w;
+	int ret;
+
+	w = sof_ipc4_find_swidget_by_ids(sdev, SOF_IPC4_MOD_ID_GET(desc->buffer_id),
+					 SOF_IPC4_MOD_INSTANCE_GET(desc->buffer_id));
```

Comment:
This is a grey area, I'm not sure if sof-clients should use swidget from sof proper.

When (If?) we move the audio cards to sof-clients then the sdev will not have a list of swidgets, all swidget will be local to the cdev, which means that you will have hard time to get this information

I think we should have a wrapper here for clients like we have for looking up the fw_module.
sof_client_ipc4_find_swidget_by_id()

The find_module() is a clean cut as the modules will always be managed by the sof core, the sof_client_ipc4_find_swidget_by_id() can be extended to search in sof-clients when the time comes and avoids using `sdev` directly.

Reply:
Ok. If I write the sof_client_ipc4_find_swidget_by_id() wrapper in sof-client.c, and maybe a warning comment above, that don't build critical features on top of this function, then this is Ok 😅?

Probes are anyway a debug feature, and temporarily breaking it at some point in a future, is not catastrophic. Still I think this feature is very helpful if someone is really trying to use probes for someting, as at least I have no idea how find out a specific module- and instance-id of a specific probe point in the topology, without going through the widgets.

Reply:
> Ok. If I write the sof_client_ipc4_find_swidget_by_id() wrapper in sof-client.c, and maybe a warning comment above, that don't build critical features on top of this function, then this is Ok 😅?

I would not add any comment, not sure  what it can help. You can only build the ipc4 variant of probes when IPC4 is enabled, so it will work.
 
> Probes are anyway a debug feature, and temporarily breaking it at some point in a future, is not catastrophic. Still I think this feature is very helpful if someone is really trying to use probes for someting, as at least I have no idea how find out a specific module- and instance-id of a specific probe point in the topology, without going through the widgets.

Right, but it is a sof-client at the same time, it has to obey the rules set for such drivers. Debug, developer or user facing modules alike.